### DISCUSSION TITLE
remark-toc Being Too Agressive

### DISCUSSION THREAD

**Discussion by edwinvelez** [remarkjs/remark#1113] [Q&A] [answered]
I am not sure if the below issue is by design or if it is bug. Please see [Code Sandobx](https://codesandbox.io/p/sandbox/pensive-http-mr896i) for a working example of the issue.

## The Issue
It seems that `remark-toc` requires the `## Table of Contents` heading be immediately followed by any other `## Subheading`. If instead it is following by a `paragraph`, the output of that paragraph is removed from output. Is this by design?  A working example of the issue can be seen in `example.md` and `example2.md`  of this [Code Sandobx](https://codesandbox.io/p/sandbox/pensive-http-mr896i) project.

## Is this intentional?
I wonder if this is intentional. If so, as a possible workaround, `example3.md` in the project may work. But only if this is by design. Can you please confirm? Is this intentional?

## Concern
This became an issue because I prefer to consume the `title` from frontmatter rather than hard-coding it in markdown. True. `example3.md` does seems to offer a workaround. But if intentional, it creates a cosmetic issue.

It is normal for the Table of Contents to be at the beginning of literature.  If `example3.md` is a valid solution, it breaks what is commonly accepted by placing it after the opening paragraph.

Could it be that this is a bug? According to the description of `remark-toc`, it should be searching for headings and ignore anything else, including paragraphs.

> It is assumed that headings define the structure of documents and that they can be linked to. When this plugin is used, authors can add a certain heading (say, ## Contents) to documents and this plugin will populate those sections with lists that link to all following sections.

Thanks for checking this out!

**Comment by ChristianMurphy** [comment, marked as answer]
> Is this by design?
> I wonder if this is intentional

Yes, it is intentional.
By design it looks for the TOC heading and the next heading (https://github.com/syntax-tree/mdast-util-toc/blob/82fc2edd9ceaad812f5f55598cd09dadc47c0494/lib/search.js#L97-L112) and replaces everything between (https://github.com/remarkjs/remark-toc/blob/608d9b151f34ab3a04769a247d71b75fdf3719cd/index.js#L31-L35)
Having multiple or different strategies for what to replace would be a feature request, and a feature request that would require some discussion on if it makes sense in TOC and what the API would look like.

> it should be searching for headings and ignore anything else, including paragraphs.

This refers to what gets added to the TOC (only headings) not replacement.

**Reply by edwinvelez** [reply]
Okay. Thank you for the confirmation.
